central/auth/datastore/singleton.go

```diff
@@ -15,7 +24,26 @@ var (
 // Singleton provides a singleton auth machine to machine DataStore.
 func Singleton() DataStore {
 	once.Do(func() {
-		ds = New(pgStore.New(globaldb.GetPostgres()))
+		set := m2m.TokenExchangerSetSingleton(roleDataStore.Singleton(), jwt.IssuerFactorySingleton())
+		ds = New(pgStore.New(globaldb.GetPostgres()), set)
+
+		// On initialization of the store, list all existing configs and fill the set.
+		ctx := sac.WithGlobalAccessScopeChecker(context.Background(), sac.AllowFixedScopes(
+			sac.AccessModeScopeKeys(storage.Access_READ_ACCESS), sac.ResourceScopeKeys(resources.Access)))
+		createTokenExchangersForExistingConfigs(ctx, ds, set)
 	})
 	return ds
 }
+
+func createTokenExchangersForExistingConfigs(ctx context.Context, ds DataStore, set m2m.TokenExchangerSet) {
+	configs, err := ds.ListAuthM2MConfigs(ctx)
+	utils.Should(err)
+	for _, config := range configs {
+		exchanger, err := set.NewTokenExchangerFromConfig(ctx, config)
+		utils.Should(err)
+		if err != nil {
+			continue
+		}
+		utils.Should(set.UpsertTokenExchanger(exchanger, config.GetId()))
+	}
+}
```

Comment:
`utils.Should(...)` usage in this function feels like a code smell to me. That's because `utils.Should()` creates a side-effect (panicking in dev builds) in a function with the intended primary effect of initializing token exchangers.

How about modifying `createTokenExchangersForExistingConfigs` to only log and return `error` and do `utils.Should()` in the place that calls it, i.e. in `func Singleton()`?

I imagine `createTokenExchangersForExistingConfigs` could look like this (untested pseudo-Go).

```go
func createTokenExchangersForExistingConfigs(ctx context.Context, ds DataStore, set m2m.TokenExchangerSet) error {
	configs, err := ds.ListAuthM2MConfigs(ctx)
    if err != nil {
        return errors.Wrap(err, "could not read m2m configs from database")
    }

    var resultingErr error
    exchangersErr := errors.New("There were errors initializing token exchangers, see log for details")

    for _, config := range configs {
		exchanger, err := set.NewTokenExchangerFromConfig(ctx, config)
		if err != nil {
            log.Warnf("Could not instantiate token exchanger %s: %v", config.GetID(), err)
            resultingErr = exchangersErr
			continue
		}
		err = set.UpsertTokenExchanger(exchanger, config.GetId())
        if err != nil {
            log.Warnf("Could not upsert token exchanger %s: %v", config.GetID(), err)
			resultingErr = exchangersErr
            continue
        }
	}

    return resultingErr
}
```

----

Separately, http calls when creating oidc providers can take up to the timeout to complete. Do you know the timeout value and retries policy there? More providers - longer Central startup time. I understand it's much easier to initialize everything before Central accepts API request because otherwise we have to think how to initialize and deal with updates concurrently, but I'm afraid we have to initialize in the background as we shouldn't make Central startup arbitrarily long.

Reply:
I've refactored the original function to move the utils.Should out.

On your part of making this asynchronous:
I moved the function to the datastore, have the singleton start a go routine which will be initializing things in the background while holding the lock for the datastore.
This seemed to me like the best solution for the time being, ensuring that the initialization function holds the lock.

Reply:
Sounds promising. I need to review now.